sirepo/auth/__init__.py

```diff
@@ -78,746 +79,649 @@
-        r = user_registration(u)
-        if cookie.unchecked_get_value(_COOKIE_METHOD) == METHOD_GUEST:
-            assert (
-                name is None
-            ), "Cookie method is {} and name is {}. Expected name to be None".format(
-                METHOD_GUEST, name
-            )
-        r.display_name = name
-        r.save()
-    cookie.set_value(_COOKIE_STATE, _STATE_LOGGED_IN)
-
-
-def create_new_user(uid_generated_callback, module):
-    import sirepo.simulation_db
-
-    u = sirepo.simulation_db.user_create()
-    uid_generated_callback(u)
-    _create_roles_for_new_user(u, module.AUTH_METHOD)
-    return u
-
+    # import simulation_db
+    sirepo.util.setattr_imports(imports)
+    _cfg = pkconfig.init(
+        methods=((METHOD_GUEST,), set, "for logging in"),
+        deprecated_methods=(set(), set, "for migrating to methods"),
+        logged_in_user=(None, str, "Only for sirepo.job_supervisor"),
+    )
+    if _cfg.logged_in_user:
+        _cfg.deprecated_methods = frozenset()
+        _cfg.methods = frozenset((METHOD_GUEST,))
+    else:
+        _init_full()
 
-def guest_uids():
-    """All of the uids corresponding to guest users."""
-    return auth_db.UserRegistration.search_all_for_column("uid", display_name=None)
 
+def only_for_api_method_modules():
+    return list(_METHOD_MODULES.values())
+
+
+class _Auth(sirepo.quest.Attr):
+    METHOD_EMAIL = METHOD_EMAIL
+    METHOD_GUEST = METHOD_GUEST
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._logged_in_user = _cfg.logged_in_user
+        self._logged_in_method = METHOD_GUEST if _cfg.logged_in_user else None
+
+    def check_sim_type_role(self, sim_type):
+        from sirepo import auth_role_moderation, oauth, uri_router
+
+        t = sirepo.template.assert_sim_type(sim_type)
+        if t not in sirepo.feature_config.auth_controlled_sim_types():
+            return
+        if not uri_router.maybe_sim_type_required_for_api(self.qcall):
+            return
+        u = self.logged_in_user()
+        r = sirepo.auth_role.for_sim_type(t)
+        if sirepo.auth_db.UserRole.has_role(
+            u, r
+        ) and not sirepo.auth_db.UserRole.is_expired(u, r):
+            return
+        elif r in sirepo.auth_role.for_proprietary_oauth_sim_types():
+            oauth.raise_authorize_redirect(self.qcall, sirepo.auth_role.sim_type(r))
+        if r in sirepo.auth_role.for_moderated_sim_types():
+            auth_role_moderation.raise_control_for_user(self.qcall, u, r)
+        sirepo.util.raise_forbidden(f"uid={u} does not have access to sim_type={t}")
+
+    def complete_registration(self, name=None):
+        """Update the database with the user's display_name and sets state to logged-in.
+        Guests will have no name.
+        """
+        u = self._qcall_bound_user()
+        with sirepo.util.THREAD_LOCK:
+            r = self.user_registration(u)
+            if self._qcall_bound_method() == METHOD_GUEST:
```

Comment:
It's a private method so I didn't think hard. I needed something to tell it that it was the method bound to the qcall not to the cookie. Under the covers it might get it from the cookie.

Reply:
8723315a3